Automattic/wp-calypso
# Multiple build targets

Implement support for multiple build targets in Calypso.

We currently have a single build, which includes everything needed to support the lowest denominator in features across supported browsers. This results in a large amount of polyfills and syntactic transformations which aren't needed for the majority of users (who thankfully do use modern browsers), and ends up bloating the bundle and increasing browser download and execution times.

This PR aims to generate a separate evergreen build in production, and serve that build instead to evergreen browsers, by doing user agent inspection. Other browsers will still work, since they have the fallback bundle, but known modern browsers will get the benefit of a smaller bundle.

There's also a small "mustard test" that tries to prevent an old browser that somehow got the modern bundle from breaking (which shouldn't happen unless it's reporting a modern browser's UA). It does this by trying to determine in runtime if the browser "cuts the mustard" by checking for support for ES modules (a reasonable test for browser "modernness" at this point in time). If it fails, the browser gets redirected to a forced fallback build with `?forceFallback=1`. This is, as @blowery put it, a "belt and suspenders" approach, in trying to ensure that there is no breakage even if we somehow detect the UA incorrectly.

**Note:** This PR replaces #30420.

#### Changes proposed in this Pull Request

* Remove duplicate browserslist definition in Babel config.
* Use the browserslist definition to configure Terser ES6 and browser-specific settings.
* Add second browserslist definition to `package.json`, for evergreen browsers.
* Modify WebPack to output to different directories based on browserslist env.
* Modify npm scripts to run two builds in production (one for fallback, one for evergreen).
* Modify server-side code to read UA and choose a bundle accordingly.
* Add "mustard test" to HTML sent to evergreen browsers.

#### Testing instructions

Evergreen:
* Open in an evergreen browser (check browserslist definition for details of which browsers that means)
* Check that page resources are coming from `/calypso/evergreen/`
* Ensure that there is no breakage when using Calypso normally

Non-evergreen:
* Open in a non-evergreen but supported browser (e.g. IE 11).
* Check that page resources are coming from `/calypso/fallback` (not `/calypso/evergreen/`)
* Ensure that there is no breakage when using Calypso normally

Compare the two:
* Compare resource transfer sizes across both versions (particularly JS). The resources sent to the evergreen browser should be smaller and include ES6 syntax such as arrow functions (`=>`).

Forced fallback:
* Open any page with `?forceFallback=1` in an evergreen browser
* Check that page resources are coming from `/calypso/fallback` (not `/calypso/evergreen/`)
* Ensure that there is no breakage when using Calypso normally
